### ISSUE TITLE
image-centric plugins should be moved under a single plugin module/folder

### ISSUE THREAD

**Issue by endofline** [hangoutsbot/hangoutsbot#753]
**UPDATE: this proposal is outdated, please scroll down to the bottom for the latest information**

proposal for new file/folder structure:

```
images
  image_links 
  image_linker_reddit -> reddit_image_linker
  image_screenshot -> selenium_screenshot
  image_memegenerator -> memegenerator

config.json: all images plugins
  images
  images.image_links
  images.reddit_image_linker
  images.selenium_screenshot
  images.memegenerator
```
* we can toss `register_shared` for the validator in https://github.com/hangoutsbot/hangoutsbot/pull/749#issuecomment-281537911 into the `images` module `__init__.py` 
* the standard to evaluate whether a plugin is migrated into the images module is as follows: does the plugin implement standalone functionality without the images module loaded? hence, plugins such as `autoreply` and `humor_hangoutcalls` are left as standalone modules 

if we didn't care too much about backward compatibility, here is an alternate structure (spoiler: I renamed `image_links`):
```
images
  image_links -> autoload
  image_linker_reddit -> reddit_image_linker
  image_screenshot -> selenium_screenshot
  image_memegenerator -> memegenerator

config.json: all images plugins
  images
  images.autoload
  images.reddit_image_linker
  images.selenium_screenshot
  images.memegenerator
```

trivia: this harks back all the way to #189! ;P

**Comment by thannemann** [comment]
My only question by doing this are we breaking backwards compatibility with current bots by moving the files and changing the config.json syntax?  I understand what you are trying to do just wanted to make sure this is a good way forward.  or should this be done in the 3.0 branch.  

And just so I understand if someone only included "images" in the config.json will this load all plugins in the directory or do they need to specifically call out all the plugins also?

**Comment by endofline** [comment]
@thannemann - both strategies outlined above will require a degree of manual reconfiguration for users if they were implemented, which is why I've opened this issue for further discussion. 

Anything for this issue should **not** be combined with the PR in #749, I have linked it as an FYI and as part of the reasoning why using `register_shared` is the best option as it "flattens" the plugin space and eliminates the fragile `from module.submodule import *` syntax. 

For the second part of your comment, loading 'images' will only call its own initialisation module - the submodules must be independently added into the configuration for them to be loaded.

In any event #749 should be committed to staging first before we make any changes - and we have yet to determine a schedule if/when this issue will be resolved. At the end of the day, we'll have to decide how much to break, and how much we can help either in the form of notifying bot devs and/or applying auto-migration functionality e.g. detecting obsolete "image_linker" plugin, then modifying config.json on the fly to add in the images plugin and so on.
